frontend/src/components/widgets/Form/Form.tsx

```diff
@@ -15,19 +15,99 @@
  * limitations under the License.
  */
 
-import React, { PureComponent, ReactNode } from "react"
+import React, { PureComponent, ReactElement, ReactNode } from "react"
+import Alert from "src/components/elements/Alert"
+import { Kind } from "src/components/shared/AlertContainer"
+import { Timer } from "src/lib/Timer"
 import { StyledForm } from "./styled-components"
 
-interface Props {
+export interface Props {
+  formId: string
   width: number
+  hasSubmitButton: boolean
 }
 
-export class Form extends PureComponent<Props> {
+interface State {
+  // Set to true when our 'submitButtonWarningTimer' expires and we
+  // don't have a submit button.
+  submitButtonTimeout: boolean
+}
+
+/**
+ * If the form doesn't have a submit button after this many milliseconds
+ * have elapsed, we show a warning that tells them they probably want to
+ * create one.
+ */
+export const SUBMIT_BUTTON_WARNING_TIME_MS = 1500
+
+export const MISSING_SUBMIT_BUTTON_WARNING =
+  "**Missing Submit Button**" +
+  "\n\nThis form has no submit button, which means that user interactions will " +
+  "never be sent to your Streamlit app." +
+  "\n\nTo create a submit button, use the `st.beta_submit_button()` function."
+
+export class Form extends PureComponent<Props, State> {
```

Comment:
Any reason why we didn't use hooks here?

Reply:
There's a long thread in Slack, but basically: `useEffect` cleanup callbacks aren't being called in Enzyme tests when a component is unmounted. These classes originally used hooks, and I rewrote them so that I could have running tests.

(Also, Enzyme's `shallow()` function is incompatible with hooks.)